library/Network/Wai/Cli.hs

```diff
@@ -82,14 +89,15 @@ runActivated run warps app = do
         setNonBlockingFD (S.fdSocket sock) True
         forkIO $ run warps sock app
     Nothing → putStrLn "No sockets to activate"
+#endif
 
 runGraceful ∷ GracefulMode → (Settings → Application → IO ()) → Settings → Application → IO ()
 runGraceful mode run warps app = do
   -- based on https://gist.github.com/NathanHowell/5435345
   -- XXX: an option to stop accepting (need change stuff inside Warp?)
   shutdown ← newEmptyTMVarIO
   activeConnections ← newTVarIO (0 ∷ Int)
-  _ ← installHandler sigTERM (CatchOnce $ atomically $ putTMVar shutdown ()) Nothing
+  _ ← installHandler sigTERM (\_ → void . atomically $ tryPutTMVar shutdown ())
```

Comment:
Here is where I'm not sure about signal handling. It was doing the `CatchOnce` (which I think I saw isn't actually guaranteed to be exactly once in the source code), but now is not doing `CatchOnce`. I changed the `putTMVar` to a `tryPutTMVar` so that if it does happen twice, at least the second one doesn't get stuck.

Reply:
The signal library doesn't look very good to me.

https://github.com/pmlodawski/signal/blob/master/src/System/Signal.hs

Hardcoded signal numbers (Linux ones??) on all platforms instead of importing from posix, does nothing other than call posix with `CatchInfo` when not on mingw.

How about just ifdef-ing out the graceful mode completely for Windows?

Reply:
Yeah, I put this back how it was and ifdef-ed out all of the *graceful* stuff for windows. [Turns out](http://neilmitchell.blogspot.com/2015/05/handling-control-c-in-haskell.html) handling ctrl-c in windows isn't so simple. I guess that can be a separate thing if required. (I'm intending on developing in windows and deploying in linux/docker, so I doubt it'll be relevant).